### PR TITLE
Pod running on edge node has invalid serviceaccount token

### LINKED ISSUE #3798: Pod running on edge node has invalid serviceaccount token
**What happened**:
Pod running on edge node has invalid serviceaccount token.

**What you expected to happen**:
Valid serviceaccount token.

**How to reproduce it (as minimally and precisely as possible)**:
Create a deploy and delete pod, After new pod is created, run `kubectl exec -it <POD_NAME>  -- cat /var/run/secrets/kubernetes.io/serviceaccount/token` get token then decode it, will find out that token's pod name field is not match current pod. Token seems come from some deleted pod, which make it invalid to use.

**Anything else we need to know?**:
I am trying to access api server from edge pod.

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.4", GitCommit:"3cce4a82b44f032d0cd1a1790e6d2f5a55d20aae", GitTreeState:"clean", BuildDate:"2021-08-11T18:16:05Z", GoVersion:"go1.16.7", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.4", GitCommit:"3cce4a82b44f032d0cd1a1790e6d2f5a55d20aae", GitTreeState:"clean", BuildDate:"2021-08-11T18:10:22Z", GoVersion:"go1.16.7", Compiler:"gc", Platform:"linux/amd64"}
```
- KubeEdge version(e.g. `cloudcore --version` and `edgecore --version`):
```
# cloudcore --version
KubeEdge v1.8.2
# edgecore --version
KubeEdge v1.8.1
```
- <details><summary>Cloud nodes Environment:</summary>

  - Hardware configuration (e.g. `lscpu`):
```
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                4
On-line CPU(s) list:   0-3
Thread(s) per core:    2
Core(s) per socket:    2
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 106
Model name:            Intel(R) Xeon(R) Platinum 8369B CPU @ 2.70GHz
Stepping:              6
CPU MHz:               2699.998
BogoMIPS:              5399.99
Hypervisor vendor:     KVM
Virtualization type:   full
L1d cache:             48K
L1i cache:             32K
L2 cache:              1280K
L3 cache:              49152K
NUMA node0 CPU(s):     0-3
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc eagerfpu pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 arat avx512vbmi pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq arch_capabilities
```

  - OS (e.g. `cat /etc/os-release`):
```
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```
  - Kernel (e.g. `uname -a`):
```
Linux master-0 3.10.0-1160.59.1.el7.x86_64 #1 SMP Wed Feb 23 16:47:03 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```
  - Go version (e.g. `go version`):
  - Others:

  </details>
- <details><summary>Edge nodes Environment:</summary>

  - edgecore version (e.g. `edgecore --version`):
```
KubeEdge v1.8.1
```
  - Hardware configuration (e.g. `lscpu`):
```
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                4
On-line CPU(s) list:   0-3
Thread(s) per core:    2
Core(s) per socket:    2
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel(R) Xeon(R) Platinum 8163 CPU @ 2.50GHz
Stepping:              4
CPU MHz:               2500.002
BogoMIPS:              5000.00
Hypervisor vendor:     KVM
Virtualization type:   full
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
L3 cache:              33792K
NUMA node0 CPU(s):     0-3
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc eagerfpu pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 arat
```
  - OS (e.g. `cat /etc/os-release`):
```
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```
  - Kernel (e.g. `uname -a`):
```
Linux worker-0 3.10.0-1160.59.1.el7.x86_64 #1 SMP Wed Feb 23 16:47:03 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```
  - Go version (e.g. `go version`):
  - Others:

  </details>

### PR DESCRIPTION
**Pull request by vincentgoat**
Signed-off-by: vincentgoat <linguohui1@huawei.com>

**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
We currently only create one serviceaccount token for one namespace in the edge node, it was saved in DB with primary KEY format `<namespace>/<restype></serviceaccount name>`, so all pods in one node use the same one serviceaccount token (which belongs to the first pod created in the edge node). It will authenticate failed after the first pod was deleted.
cc @fisherxu @kevin-wangzefeng 

**Which issue(s) this PR fixes**:

Fixes #3798

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by vincentgoat** [comment]
cc @fisherxu @wackxu @gy95 PTAL.

File: edge/pkg/metamanager/process.go [hunk 1]
```diff
@@ -238,11 +239,30 @@ func (m *metaManager) processDelete(message model.Message) {
 	sendToCloud(resp)
 }
 
+// KeyFunc keys should be nonconfidential and safe to log
+func KeyFunc(resource string, tr *authenticationv1.TokenRequest) string {
+	return fmt.Sprintf("%s/%s/%s/%s", resource, tr.Spec.BoundObjectRef.Kind, tr.Spec.BoundObjectRef.Name, tr.Spec.BoundObjectRef.UID)
```
Comment (wackxu) [inline edge/pkg/metamanager/process.go:244]: In kubernetes， the key is
```
func keyFunc(name, namespace string, tr *authenticationv1.TokenRequest) string {
	var exp int64
	if tr.Spec.ExpirationSeconds != nil {
		exp = *tr.Spec.ExpirationSeconds
	}

	var ref authenticationv1.BoundObjectReference
	if tr.Spec.BoundObjectRef != nil {
		ref = *tr.Spec.BoundObjectRef
	}

	return fmt.Sprintf("%q/%q/%#v/%#v/%#v", name, namespace, tr.Spec.Audiences, exp, ref)
}
```

if user update the token request info in pod, the token is different
```
  volumes:
  - name: istio-token
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          audience: istio-ca
          expirationSeconds: 43200
          path: istio-token
```

so we should consider the key func to adapt to these cases

File: edge/pkg/metamanager/client/serviceaccount.go [hunk 2]
```diff
@@ -21,20 +26,103 @@ type ServiceAccountTokenGetter interface {
...
+	}
+	// Require a refresh if within 20% of the TTL from the expiration time.
+	if now.After(exp.Add(-1 * time.Duration((*tr.Spec.ExpirationSeconds*20)/100) * time.Second)) {
+		return true
+	}
+	return false
+}
+
+func getTokenLocally(resource string, tr *authenticationv1.TokenRequest) (*authenticationv1.TokenRequest, error) {
+	resKey := metamanager.KeyFunc(resource, tr)
```
Comment (wackxu) [inline edge/pkg/metamanager/client/serviceaccount.go:97]: The token is stored in plaintext, we should consider its security

File: edge/pkg/metamanager/client/serviceaccount.go [hunk 3]
```diff
@@ -21,20 +26,103 @@ type ServiceAccountTokenGetter interface {
...
+	switch len(tokens) {
+	case 6:
+		return types.UID(tokens[len(tokens)-1])
+	default:
+	}
+	return ""
+}
+
+func (c *serviceAccountToken) DeleteServiceAccountToken(podUID types.UID) {
+	svcAccounts, err := dao.QueryAllMeta("type", model.ResourceTypeServiceAccountToken)
```
Comment (wackxu) [inline edge/pkg/metamanager/client/serviceaccount.go:57]: QueryAllMeta performance may be relatively low, could we use sql like to query

File: edge/pkg/metamanager/client/serviceaccount.go [hunk 4]
```diff
@@ -21,20 +26,103 @@ type ServiceAccountTokenGetter interface {
...
+	// Require a refresh if within 20% of the TTL from the expiration time.
+	if now.After(exp.Add(-1 * time.Duration((*tr.Spec.ExpirationSeconds*20)/100) * time.Second)) {
+		return true
+	}
+	return false
+}
+
+func getTokenLocally(resource string, tr *authenticationv1.TokenRequest) (*authenticationv1.TokenRequest, error) {
+	resKey := metamanager.KeyFunc(resource, tr)
+	metas, err := dao.QueryMeta("key", resKey)
```
Comment (wackxu) [inline edge/pkg/metamanager/client/serviceaccount.go:98]: ditto

File: edge/pkg/metamanager/client/serviceaccount.go [hunk 5]
```diff
@@ -54,6 +142,15 @@ func (c *serviceAccountToken) GetServiceAccountToken(namespace string, name stri
 	return handleServiceAccountTokenFromMetaManager(content)
 }
 
+func (c *serviceAccountToken) GetServiceAccountToken(namespace string, name string, tr *authenticationv1.TokenRequest) (*authenticationv1.TokenRequest, error) {
+	resource := fmt.Sprintf("%s/%s/%s", namespace, model.ResourceTypeServiceAccountToken, name)
+	tokenReq, err := getTokenLocally(resource, tr)
```
Comment (wackxu) [inline edge/pkg/metamanager/client/serviceaccount.go:147]: for service account token、 secret、configmap， the volume plugin will periodiclly query the resource info, so we need a cache manager to reduce the query from sqlite.

Comment (wackxu) [reply edge/pkg/metamanager/client/serviceaccount.go:147]: may be we can add a issue first to trace it

File: edge/pkg/metamanager/process.go (hunk 1 above)
Comment (vincentgoat) [reply edge/pkg/metamanager/process.go]: `return fmt.Sprintf("%q/%#v/%#v/%#v", resource, tr.Spec.Audiences, exp, ref.UID)`
The max length of `key` in sqlite db is  256, minimize the `key` length by store `ref.UID`.

File: edge/pkg/metamanager/client/serviceaccount.go (hunk 5 above)
Comment (vincentgoat) [reply edge/pkg/metamanager/client/serviceaccount.go:147]: Yes, Traced by https://github.com/kubeedge/kubeedge/issues/4144

File: edge/pkg/metamanager/client/serviceaccount.go (hunk 3 above)
Comment (vincentgoat) [reply edge/pkg/metamanager/client/serviceaccount.go:57]: New issue to trace.

File: edge/pkg/metamanager/client/serviceaccount.go (hunk 2 above)
Comment (vincentgoat) [reply edge/pkg/metamanager/client/serviceaccount.go:97]: Issue https://github.com/kubeedge/kubeedge/issues/4014 will trace this problem, and we will fix it later.

**Comment by fisherxu** [comment]
When we merge the new edged, does the fix still needed? cc @Shelley-BaoYue @vincentgoat @wackxu

**Comment by Shelley-BaoYue** [comment]
> When we merge the new edged, does the fix still needed? cc @Shelley-BaoYue @vincentgoat @wackxu

I think it's needed. And `DeleteServiceAccountFunc` in Kubelet only deletes cache info, we also need to make some modification in Kubelet to support delete service account saved in DB at the same time.

**Comment by vincentgoat** [comment]
When we merged the new Edged, we won't need this change. 
However, after we merged the new Edged,
1. we need to maintain the `serviceaccounttoken` deletion in the SQLite DB.
2. update the `keyFunc` in the SQLite DB.

Besides, we can merge this PR into the latest 3 minor versions.
@Shelley-BaoYue @fisherxu @wackxu

**Comment by Shelley-BaoYue** [comment]
In new edged, we can add DeleteServiceAccount in Kubelet. FYI: https://github.com/kubeedge/kubernetes/pull/1/commits/ceedd0f45997c09b779e9474dffdd6362b0368fb 

@fisherxu @vincentgoat

**Review by fisherxu — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**